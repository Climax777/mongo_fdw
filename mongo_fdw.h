/*-------------------------------------------------------------------------
 *
 * mongo_fdw.h
 *
 * Type and function declarations for MongoDB foreign data wrapper.
 *
 * Copyright (c) 2012-2014 Citus Data, Inc.
 *
 *-------------------------------------------------------------------------
 */

#ifndef MONGO_FDW_H
#define MONGO_FDW_H

#include "bson.h"
#include "mongo.h"

#include "postgres.h"
#include "utils/hsearch.h"
#include "fmgr.h"
#include "catalog/pg_foreign_server.h"
#include "catalog/pg_foreign_table.h"
#include "utils/datetime.h"
#include "nodes/pg_list.h"
#include "nodes/relation.h"
#include "utils/timestamp.h"
#include "access/reloptions.h"
#include "catalog/pg_type.h"
#include "commands/defrem.h"
#include "commands/explain.h"
#include "commands/vacuum.h"
#include "foreign/fdwapi.h"
#include "foreign/foreign.h"
#include "nodes/makefuncs.h"
#include "optimizer/cost.h"
#include "optimizer/pathnode.h"
#include "optimizer/plancat.h"
#include "optimizer/planmain.h"
#include "optimizer/restrictinfo.h"
#include "utils/array.h"
#include "utils/builtins.h"
#include "utils/date.h"
#include "utils/hsearch.h"
#include "utils/lsyscache.h"
#include "utils/rel.h"
#include "utils/memutils.h"



/* Defines for valid option names */
#define OPTION_NAME_ADDRESS "address"
#define OPTION_NAME_PORT "port"
#define OPTION_NAME_DATABASE "database"
#define OPTION_NAME_COLLECTION "collection"

/* Default values for option parameters */
#define DEFAULT_IP_ADDRESS "127.0.0.1"
#define DEFAULT_PORT_NUMBER 27017
#define DEFAULT_DATABASE_NAME "test"

/* Defines for sending queries and converting types */
#define EQUALITY_OPERATOR_NAME "="
#define INITIAL_ARRAY_CAPACITY 8
#define MONGO_TUPLE_COST_MULTIPLIER 5
#define MONGO_CONNECTION_COST_MULTIPLIER 5
#define POSTGRES_TO_UNIX_EPOCH_DAYS (POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE)
#define POSTGRES_TO_UNIX_EPOCH_USECS (POSTGRES_TO_UNIX_EPOCH_DAYS * USECS_PER_DAY)


/*
 * MongoValidOption keeps an option name and a context. When an option is passed
 * into mongo_fdw objects (server and foreign table), we compare this option's
 * name and context against those of valid options.
 */
typedef struct MongoValidOption
{
	const char *optionName;
	Oid optionContextId;

} MongoValidOption;


/* Array of options that are valid for mongo_fdw */
static const uint32 ValidOptionCount = 4;
static const MongoValidOption ValidOptionArray[] =
{
	/* foreign server options */
	{ OPTION_NAME_ADDRESS, ForeignServerRelationId },
	{ OPTION_NAME_PORT,  ForeignServerRelationId },

	/* foreign table options */
	{ OPTION_NAME_DATABASE, ForeignTableRelationId },
	{ OPTION_NAME_COLLECTION, ForeignTableRelationId }
};


/*
 * MongoFdwOptions holds the option values to be used when connecting to Mongo.
 * To resolve these values, we first check foreign table's options, and if not
 * present, we then fall back to the default values specified above.
 */
typedef struct MongoFdwOptions
{
	char *addressName;
	int32 portNumber;
	char *databaseName;
	char *collectionName;

} MongoFdwOptions;


/*
* MongoFdwExecState keeps foreign data wrapper specific execution state that we
* create and hold onto when executing the query.
*/
/*
* Execution state of a foreign insert/update/delete operation.
*/
typedef struct MongoFdwModifyState
{
	Relation		rel;				/* relcache entry for the foreign table */

	List			*target_attrs;		/* list of target attribute numbers */

	/* info about parameters for prepared statement */
	int				p_nums;				/* number of parameters to transmit */
	FmgrInfo		*p_flinfo;			/* output conversion functions for them */

	struct HTAB 	*columnMappingHash;

	mongo			*mongoConnection;	/* MongoDB connection */
	mongo_cursor	*mongoCursor;		/* MongoDB cursor */
	bson			*queryDocument;		/* Bson Document */

	MongoFdwOptions *mongoFdwOptions;

	/* working memory context */
	MemoryContext temp_cxt;         /* context for per-tuple temporary data */
} MongoFdwModifyState;


/*
 * ColumnMapping reprents a hash table entry that maps a column name to column
 * related information. We construct these hash table entries to speed up the
 * conversion from BSON documents to PostgreSQL tuples; and each hash entry maps
 * the column name to the column's tuple index and its type-related information.
 */
typedef struct ColumnMapping
{
	char columnName[NAMEDATALEN];
	uint32 columnIndex;
	Oid columnTypeId;
	int32 columnTypeMod;
	Oid columnArrayTypeId;
} ColumnMapping;


/* Function declarations related to creating the mongo query */
extern List * ApplicableOpExpressionList(RelOptInfo *baserel);
extern bson * QueryDocument(Oid relationId, List *opExpressionList);
extern List * ColumnList(RelOptInfo *baserel);

/* Function declarations for foreign data wrapper */
extern Datum mongo_fdw_handler(PG_FUNCTION_ARGS);
extern Datum mongo_fdw_validator(PG_FUNCTION_ARGS);
extern MongoFdwOptions * MongoGetOptions(Oid foreignTableId);
extern void MongoFreeOptions(MongoFdwOptions *mongoFdwOptions);

mongo* GetConnection(char *host, int32 port);
void cleanup_connection(void);
void ReleaseConnection(mongo *conn);

#endif   /* MONGO_FDW_H */
