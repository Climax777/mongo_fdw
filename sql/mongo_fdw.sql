\c postgres postgres
CREATE EXTENSION mongo_fdw;
CREATE SERVER mongo_server FOREIGN DATA WRAPPER mongo_fdw OPTIONS (address '127.0.0.1', port '27017');
\! mongoimport --db mongo_fdw_regress --collection countries --jsonArray --drop --maintainInsertionOrder --host='127.0.0.1' --port=27017 --quiet < data/mongo_fixture.json
CREATE USER MAPPING FOR postgres SERVER mongo_server;

CREATE FOREIGN TABLE department(_id NAME, department_id int, department_name text) SERVER mongo_server OPTIONS(database 'testdb', collection 'department');
CREATE FOREIGN TABLE employee(_id NAME, emp_id int, emp_name text, emp_dept_id int) SERVER mongo_server OPTIONS(database 'testdb', collection 'employee');

INSERT INTO department SELECT 0, i, 'dept - ' || i FROM generate_series(1,10) i;
INSERT INTO employee SELECT 0, i, 'emp - ' || i, (i - 1)%10 + 1 FROM generate_series(1,100) i;

SELECT count(*) FROM department;
SELECT count(*) FROM employee;

EXPLAIN (COSTS FALSE) SELECT emp_id , emp_name , emp_dept_id, department_id , department_name  FROM department d, employee e WHERE d.department_id = e.emp_dept_id ORDER by emp_id;

EXPLAIN (COSTS FALSE) SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id IN (SELECT department_id FROM department) ORDER by emp_id;

SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id = e.emp_dept_id AND e.emp_dept_id > 5 ORDER by emp_id, department_id;
SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id IN (SELECT department_id FROM department WHERE department_id < 3) ORDER by emp_id, department_id;

DELETE FROM employee WHERE emp_id = 10;

UPDATE employee SET emp_name = 'Updated emp' WHERE emp_id = 20;
SELECT emp_id, emp_name FROM employee WHERE emp_name like 'Updated emp' ORDER BY emp_id;

SELECT emp_id , emp_name , emp_dept_id FROM employee ORDER by emp_id LIMIT 10;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (1) ORDER by emp_id;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (1,3,4,5) ORDER by emp_id;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (10000,1000) ORDER by emp_id;

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (1)  ORDER by emp_id LIMIT 5;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (1,3,4,5) ORDER by emp_id LIMIT 5;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (10000,1000) ORDER by emp_id LIMIT 5;

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (SELECT emp_id FROM employee WHERE emp_id IN (1,10)) ORDER by emp_id;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_name NOT IN ('emp - 1', 'emp - 2') ORDER by emp_id LIMIT 5;
SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_name NOT IN ('emp - 10') ORDER by emp_id LIMIT 5;

DELETE FROM employee;
DELETE FROM department;

CREATE FOREIGN TABLE countries (
_id NAME,
name VARCHAR,
population INTEGER,
capital VARCHAR,
hdi FLOAT
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM countries ORDER BY _id;
-- 
-- Subfields and dates
CREATE FOREIGN TABLE country_elections (
_id NAME,
"lastElections.type" VARCHAR,
"lastElections.date" TIMESTAMP
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM country_elections ORDER BY _id;
-- 
-- Arrays
CREATE FOREIGN TABLE main_exports (
_id NAME,
"mainExports" TEXT[]
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM main_exports ORDER BY _id;

-- __doc tests

-- the collection warehouse must contain the following data
-- use testdb;
-- db.warehouse.insert ({"_id" : ObjectId("58a1ebbaf543ec0b90545859"),"warehouse_id" : NumberInt(1),"warehouse_name" : "UPS","warehouse_created" : ISODate("2014-12-12T07:12:10Z")});
-- db.warehouse.insert ({"_id" : ObjectId("58a1ebbaf543ec0b9054585a"),"warehouse_id" : NumberInt(2),"warehouse_name" : "Laptop","warehouse_created" : ISODate("2015-11-11T08:13:10Z")});


CREATE FOREIGN TABLE test_json(__doc json) SERVER mongo_server OPTIONS (database 'testdb', collection 'warehouse');
CREATE FOREIGN TABLE test_jsonb(__doc jsonb) SERVER mongo_server OPTIONS (database 'testdb', collection 'warehouse');
CREATE FOREIGN TABLE test_text(__doc text) SERVER mongo_server OPTIONS (database 'testdb', collection 'warehouse');
CREATE FOREIGN TABLE test_varchar(__doc varchar) SERVER mongo_server OPTIONS (database 'testdb', collection 'warehouse');

SELECT * FROM test_json ORDER BY __doc::text COLLATE "C";
SELECT * FROM test_jsonb ORDER BY __doc::text COLLATE "C";
SELECT * FROM test_text ORDER BY __doc::text COLLATE "C";
SELECT * FROM test_varchar ORDER BY __doc::text COLLATE "C";

-- where clause push down test
CREATE FOREIGN TABLE test_numbers(_id NAME, a int, b text) SERVER mongo_server OPTIONS (database 'testdb', collection 'test_numbers');
insert into test_numbers values('1', 1, 'One');
insert into test_numbers values('2', 2, 'Two');
insert into test_numbers values('3', 3, 'Three');
insert into test_numbers values('4', 4, 'Four');
insert into test_numbers values('5', 5, 'Five');
insert into test_numbers values('6', 6, 'Six');
insert into test_numbers values('7', 7, 'Seven');
insert into test_numbers values('8', 8, 'Eight');
insert into test_numbers values('9', 9, 'Nine');
insert into test_numbers values('10', 10, 'Ten');

create or replace function test_param_where() returns void as $$
DECLARE
  n varchar;
BEGIN
  FOR x IN 1..9 LOOP
    select b into n from test_numbers where a=x;
    raise notice 'Found Item %', n;
  end loop;
  return;
END
$$ LANGUAGE plpgsql;

SELECT test_param_where();

PREPARE test_where_pd(int) as SELECT b FROM test_numbers WHERE a =$1;
explain (verbose, costs false) execute test_where_pd(1);
explain (verbose, costs false) execute test_where_pd(2);
explain (verbose, costs false) execute test_where_pd(3);
explain (verbose, costs false) execute test_where_pd(4);
explain (verbose, costs false) execute test_where_pd(5);
explain (verbose, costs false) execute test_where_pd(6);
explain (verbose, costs false) execute test_where_pd(7);
explain (verbose, costs false) execute test_where_pd(8);
explain (verbose, costs false) execute test_where_pd(9);

execute test_where_pd(1);
execute test_where_pd(2);
execute test_where_pd(3);
execute test_where_pd(4);
execute test_where_pd(5);
execute test_where_pd(6);
execute test_where_pd(7);
execute test_where_pd(8);
execute test_where_pd(9);

--
-- fdw-108: After a change to a pg_foreign_server or pg_user_mapping catalog
-- entry, connection should be invalidated.
--

-- Alter one of the SERVER option
-- Set wrong address for mongo_server
ALTER SERVER mongo_server OPTIONS (SET address '127.0.0.5');
-- Should fail with an error
INSERT INTO test_numbers VALUES ('11', 11, 'Eleven');
UPDATE test_numbers SET a = 11 WHERE a = 10;
DELETE FROM test_numbers WHERE a = 10;
SELECT * FROM test_numbers;
-- Set correct address for mongo_server
ALTER SERVER mongo_server OPTIONS (SET address '127.0.0.1');
-- Should able to insert the data
INSERT INTO test_numbers VALUES ('12', 12, 'Twelve');

-- Change the user mapping options
-- Set wrong username, password for postgres user
DROP USER MAPPING FOR postgres SERVER mongo_server;
CREATE USER MAPPING FOR postgres SERVER mongo_server OPTIONS (username 'wrong', password 'wrong');
-- Should fail with an error
INSERT INTO test_numbers VALUES ('13', 13, 'Thirteen');
UPDATE test_numbers SET a = 11 WHERE a = 10;
DELETE FROM test_numbers WHERE a = 10;
SELECT * FROM test_numbers;
-- Set default username, password for postgres user
DROP USER MAPPING FOR postgres SERVER mongo_server;
CREATE USER MAPPING FOR postgres SERVER mongo_server;
-- Should able to insert the data
INSERT INTO test_numbers VALUES ('14', 14, 'Fourteen');

DELETE FROM test_numbers;
DROP FOREIGN TABLE test_numbers;

DROP FOREIGN TABLE test_json;
DROP FOREIGN TABLE test_jsonb;
DROP FOREIGN TABLE test_text;
DROP FOREIGN TABLE test_varchar;

DROP FOREIGN TABLE department;
DROP FOREIGN TABLE employee;
DROP FOREIGN TABLE countries;
DROP FOREIGN TABLE country_elections;
DROP FOREIGN TABLE main_exports;
DROP USER MAPPING FOR postgres SERVER mongo_server;
DROP EXTENSION mongo_fdw CASCADE;
