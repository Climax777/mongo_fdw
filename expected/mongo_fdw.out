CREATE SERVER mongo_server FOREIGN DATA WRAPPER mongo_fdw OPTIONS (address '127.0.0.1', port '27017');
\! mongoimport --db mongo_fdw_regress --collection countries --jsonArray --drop --quiet < data/mongo_fixture.json
CREATE USER MAPPING FOR postgres SERVER mongo_server;
CREATE FOREIGN TABLE department(_id NAME, department_id int, department_name text) SERVER mongo_server OPTIONS(database 'testdb', collection 'department');
CREATE FOREIGN TABLE employee(_id NAME, emp_id int, emp_name text, emp_dept_id int) SERVER mongo_server OPTIONS(database 'testdb', collection 'employee');
INSERT INTO department VALUES(0, generate_series(1,100), 'dept - ' || generate_series(1,100));
INSERT INTO employee VALUES(0, generate_series(1,100), 'emp - ' || generate_series(1,100), generate_series(1,100));
SELECT count(*) FROM department;
 count 
-------
   100
(1 row)

SELECT count(*) FROM employee;
 count 
-------
   100
(1 row)

EXPLAIN (COSTS FALSE) SELECT emp_id , emp_name , emp_dept_id, department_id , department_name  FROM department d, employee e WHERE d.department_id = e.emp_dept_id LIMIT 10;
                       QUERY PLAN                       
--------------------------------------------------------
 Limit
   ->  Hash Join
         Hash Cond: (d.department_id = e.emp_dept_id)
         ->  Foreign Scan on department d
               Foreign Namespace: testdb.department
         ->  Hash
               ->  Foreign Scan on employee e
                     Foreign Namespace: testdb.employee
(8 rows)

EXPLAIN (COSTS FALSE) SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id IN (SELECT department_id FROM department) LIMIT 10;
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Limit
   ->  Nested Loop
         ->  Nested Loop Semi Join
               Join Filter: (d.department_id = department.department_id)
               ->  Foreign Scan on department d
                     Foreign Namespace: testdb.department
               ->  Materialize
                     ->  Foreign Scan on department
                           Foreign Namespace: testdb.department
         ->  Materialize
               ->  Foreign Scan on employee e
                     Foreign Namespace: testdb.employee
(12 rows)

SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id = e.emp_dept_id LIMIT 10;
 emp_id | emp_name | emp_dept_id | department_id | department_name 
--------+----------+-------------+---------------+-----------------
     39 | emp - 39 |          39 |            39 | dept - 39
     40 | emp - 40 |          40 |            40 | dept - 40
     41 | emp - 41 |          41 |            41 | dept - 41
     42 | emp - 42 |          42 |            42 | dept - 42
     43 | emp - 43 |          43 |            43 | dept - 43
     44 | emp - 44 |          44 |            44 | dept - 44
     45 | emp - 45 |          45 |            45 | dept - 45
     46 | emp - 46 |          46 |            46 | dept - 46
     47 | emp - 47 |          47 |            47 | dept - 47
     48 | emp - 48 |          48 |            48 | dept - 48
(10 rows)

SELECT emp_id , emp_name , emp_dept_id, department_id , department_name FROM department d, employee e WHERE d.department_id IN (SELECT department_id FROM department) LIMIT 10;
 emp_id | emp_name | emp_dept_id | department_id | department_name 
--------+----------+-------------+---------------+-----------------
     38 | emp - 38 |          38 |            39 | dept - 39
     39 | emp - 39 |          39 |            39 | dept - 39
     40 | emp - 40 |          40 |            39 | dept - 39
     41 | emp - 41 |          41 |            39 | dept - 39
     42 | emp - 42 |          42 |            39 | dept - 39
     43 | emp - 43 |          43 |            39 | dept - 39
     44 | emp - 44 |          44 |            39 | dept - 39
     45 | emp - 45 |          45 |            39 | dept - 39
     46 | emp - 46 |          46 |            39 | dept - 39
     47 | emp - 47 |          47 |            39 | dept - 39
(10 rows)

DELETE FROM employee WHERE emp_id = 10;
UPDATE employee SET emp_name = 'Updated emp' WHERE emp_id = 20;
SELECT emp_id, emp_name FROM employee WHERE emp_name like 'Updated emp';
 emp_id |  emp_name   
--------+-------------
     20 | Updated emp
(1 row)

SELECT emp_id , emp_name , emp_dept_id FROM employee LIMIT 10;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
     43 | emp - 43 |          43
     44 | emp - 44 |          44
     45 | emp - 45 |          45
     46 | emp - 46 |          46
     47 | emp - 47 |          47
(10 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (1);
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
      1 | emp - 1  |           1
(1 row)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (1,3,4,5);
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
      1 | emp - 1  |           1
      3 | emp - 3  |           3
      4 | emp - 4  |           4
      5 | emp - 5  |           5
(4 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id IN (10000,1000);
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
(0 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (1) LIMIT 5;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
(5 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (1,3,4,5) LIMIT 5;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
(5 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (10000,1000) LIMIT 5;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
(5 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_id NOT IN (SELECT emp_id FROM employee WHERE emp_id IN (1,10));
 emp_id |  emp_name   | emp_dept_id 
--------+-------------+-------------
     38 | emp - 38    |          38
     39 | emp - 39    |          39
     40 | emp - 40    |          40
     41 | emp - 41    |          41
     42 | emp - 42    |          42
     43 | emp - 43    |          43
     44 | emp - 44    |          44
     45 | emp - 45    |          45
     46 | emp - 46    |          46
     47 | emp - 47    |          47
     48 | emp - 48    |          48
     49 | emp - 49    |          49
     50 | emp - 50    |          50
     51 | emp - 51    |          51
     52 | emp - 52    |          52
     53 | emp - 53    |          53
     54 | emp - 54    |          54
     55 | emp - 55    |          55
     56 | emp - 56    |          56
     57 | emp - 57    |          57
     58 | emp - 58    |          58
     59 | emp - 59    |          59
     60 | emp - 60    |          60
     61 | emp - 61    |          61
     62 | emp - 62    |          62
     63 | emp - 63    |          63
     64 | emp - 64    |          64
     65 | emp - 65    |          65
     66 | emp - 66    |          66
     67 | emp - 67    |          67
     68 | emp - 68    |          68
     69 | emp - 69    |          69
     70 | emp - 70    |          70
     71 | emp - 71    |          71
     72 | emp - 72    |          72
     73 | emp - 73    |          73
     74 | emp - 74    |          74
     75 | emp - 75    |          75
     76 | emp - 76    |          76
     77 | emp - 77    |          77
     78 | emp - 78    |          78
     79 | emp - 79    |          79
     80 | emp - 80    |          80
     81 | emp - 81    |          81
     82 | emp - 82    |          82
     83 | emp - 83    |          83
     84 | emp - 84    |          84
     85 | emp - 85    |          85
     86 | emp - 86    |          86
     87 | emp - 87    |          87
     88 | emp - 88    |          88
     89 | emp - 89    |          89
     90 | emp - 90    |          90
     91 | emp - 91    |          91
     92 | emp - 92    |          92
     93 | emp - 93    |          93
     94 | emp - 94    |          94
     95 | emp - 95    |          95
     96 | emp - 96    |          96
     97 | emp - 97    |          97
     98 | emp - 98    |          98
     99 | emp - 99    |          99
      2 | emp - 2     |           2
      3 | emp - 3     |           3
      4 | emp - 4     |           4
      5 | emp - 5     |           5
      6 | emp - 6     |           6
      7 | emp - 7     |           7
      8 | emp - 8     |           8
      9 | emp - 9     |           9
     11 | emp - 11    |          11
     12 | emp - 12    |          12
     13 | emp - 13    |          13
     14 | emp - 14    |          14
     15 | emp - 15    |          15
     16 | emp - 16    |          16
     17 | emp - 17    |          17
     18 | emp - 18    |          18
     19 | emp - 19    |          19
     20 | Updated emp |          20
     21 | emp - 21    |          21
     22 | emp - 22    |          22
     23 | emp - 23    |          23
     24 | emp - 24    |          24
     25 | emp - 25    |          25
     26 | emp - 26    |          26
     27 | emp - 27    |          27
     28 | emp - 28    |          28
     29 | emp - 29    |          29
     30 | emp - 30    |          30
     31 | emp - 31    |          31
     32 | emp - 32    |          32
     33 | emp - 33    |          33
     34 | emp - 34    |          34
     35 | emp - 35    |          35
     36 | emp - 36    |          36
     37 | emp - 37    |          37
    100 | emp - 100   |         100
(98 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_name NOT IN ('emp - 1', 'emp - 2') LIMIT 5;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
(5 rows)

SELECT emp_id , emp_name , emp_dept_id FROM employee WHERE emp_name NOT IN ('emp - 10') LIMIT 5;
 emp_id | emp_name | emp_dept_id 
--------+----------+-------------
     38 | emp - 38 |          38
     39 | emp - 39 |          39
     40 | emp - 40 |          40
     41 | emp - 41 |          41
     42 | emp - 42 |          42
(5 rows)

DELETE FROM employee;
DELETE FROM department;
CREATE FOREIGN TABLE countries (
_id NAME,
name VARCHAR,
population INTEGER,
capital VARCHAR,
hdi FLOAT
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM countries;
           _id            |  name   | population | capital  |  hdi  
--------------------------+---------+------------+----------+-------
 5381ccf9d6d81c8e8bf0434f | Ukraine |   45590000 | Kyiv     |  0.74
 5381ccf9d6d81c8e8bf04350 | Poland  |   38540000 | Warsaw   | 0.821
 5381ccf9d6d81c8e8bf04351 | Moldova |    3560000 | Chișinău |  0.66
(3 rows)

-- 
-- Subfields and dates
CREATE FOREIGN TABLE country_elections (
_id NAME,
"lastElections.type" VARCHAR,
"lastElections.date" TIMESTAMP
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM country_elections;
           _id            | lastElections.type |      lastElections.date      
--------------------------+--------------------+------------------------------
 5381ccf9d6d81c8e8bf0434f | presedential       | Sat Jan 10 10:51:01.504 1970
 5381ccf9d6d81c8e8bf04350 | parliamentary      | Fri Dec 26 22:44:00.128 1969
 5381ccf9d6d81c8e8bf04351 | parliamentary      | Wed Dec 10 05:00:43.904 1969
(3 rows)

-- 
-- Arrays
CREATE FOREIGN TABLE main_exports (
_id NAME,
"mainExports" TEXT[]
) SERVER mongo_server OPTIONS (database 'mongo_fdw_regress', collection 'countries');
SELECT * FROM main_exports;
           _id            |                                                                                         mainExports                                                                                         
--------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 5381ccf9d6d81c8e8bf0434f | {"Semi-finished products of iron or non-alloy steel","Flat-rolled products of iron or non-alloy steel","Sunflower-seed, safflower or cotton-seed oil"}
 5381ccf9d6d81c8e8bf04350 | {"Parts and accessories of the motor vehicles of headings 87.01 to 87.0","Motor cars and other motor vehicles principally designed for the transport","Reception apparatus for television"}
 5381ccf9d6d81c8e8bf04351 | {"Wine of fresh grapes, including fortified wines","Insulated (including enamelled or anodised) wire, cable","Sunflower seeds, whether or not broken"}
(3 rows)

DROP FOREIGN TABLE department;
DROP FOREIGN TABLE employee;
DROP FOREIGN TABLE countries;
DROP FOREIGN TABLE country_elections;
DROP FOREIGN TABLE main_exports;
DROP USER MAPPING FOR postgres SERVER mongo_server;
DROP EXTENSION mongo_fdw CASCADE;
NOTICE:  drop cascades to server mongo_server
